### PR TITLE
changes and addition to enable unittests

### PR DESCRIPTION
**Pull request by Oceancolour-RG**
In this PR the following changes were made:
1) the ability of GlintCorr to accept a `datacube.model.Dataset` or a `pathlib.Path` of the `_final.odc-metadata.yaml file`. The latter will be useful for developing unittests that uses a small dataset suite;
2) the addition of a dataset that will be used for unittests. Here **lmbadj** wagl-water-atcor products were used as water pixels in **lmbskyg** products may contain negative reflectances;
3) removed reference to `ref_band` variable in `sungc/rasterio_funcs.py:204`;
4) flake8 changes

**For commits made on 30 November**:
38 unittests were created for:
1) NIR subtraction algorithm;
2) Cox and Munk algorithm;
3) Hedley et al. (2005) algorithm;
4) Quicklook RGB generation;
5) Interactive display;
6) Different geometries in shapefiles;
7) rasterio (resample) functions, and;
8) GlintCorr class

The additions were added throughout the sunglint correction module:
a) identification of bands that only contain pixels with nodata values;
b) replacing os.path calls with pathlib.Path, and;
c) additional input checks and raising of Exceptions

Several bugs, discovered through the creation of these unittests, were corrected.

### CODE REVIEW

File: sungc/deglint.py [hunk 1]
```diff
@@ -459,14 +488,14 @@ def nir_subtraction(self, vis_band_ids, nir_band_id, odir=None, waterVal=5):
             The path where the deglinted geotiff bands are saved.
 
             if None then:
-            odir = pjoin(self.group_path, "DEGLINT", "GAO")
+            odir = pjoin(self.group_path, "DEGLINT", "MINUS_NIR")
```
Comment (sixy6e) [inline sungc/deglint.py]: Just checking that this should be "MINUS_NIR" and not "GAO"

File: sungc/rasterio_funcs.py [hunk 2]
```diff
@@ -72,20 +72,20 @@ def load_bands(band_list, scale_factor, apply_scaling):
     spectral_cube : numpy.ndarray
         multi-band array with dimensions of [nBands, nRows, nCols]
```
Comment (sixy6e) [inline sungc/rasterio_funcs.py]: If you're feeling up to it, you could update the variable names as well.  Not a hard requirement though, but it just makes it consistent.

File: sungc/deglint.py (hunk 1 above)
Comment (Oceancolour-RG) [reply sungc/deglint.py]: I changed the folder name from GAO to MINUS_NIR to better represent what the algorithm does

File: sungc/rasterio_funcs.py (hunk 2 above)
Comment (Oceancolour-RG) [reply sungc/rasterio_funcs.py]: Latest commits have updated variable names in the comments

**Review by sixy6e — APPROVED** [review]
Looks alright.  A bit more aligned with PEP standards :smiley: 

With the visualise module, there is a bunch of code dealing with Fmask.  With the data being generated for the DEA samples, it will be masked using the `MNDWI < 0`.  Just thinking that those data samples would cause an issue if then using Fmask to flag water pixels.

File: sungc/deglint.py [hunk 3]
```diff
@@ -249,37 +313,59 @@ def get_band_list(self):
...
+            Exception if any of the bands do not exist
         """
-        bandList = []
-        bandNums = []
-        bandNames = []
+        # a list of bands that are not needed for deglinting
+        skip_bands = [
+            "contiguity",
+            "azimuthal_exiting",
+            "azimuthal-incident",
```
Comment (sixy6e) [inline sungc/deglint.py]: just checking that the `-` and `_` discrepancy is correct

File: sungc/deglint.py [hunk 4]
```diff
@@ -359,35 +454,40 @@ def quicklook_RGB(self, dwnscale_factor=3):
...
         # a very pretty RGB
-        rgb_im = seadas_style_RGB(
+        rgb_im = seadas_style_rgb(
             refl_img=refl_im, rgb_ix=[0, 1, 2], scale_factor=self.scale_factor
         )
 
+        rio_meta["band_1"] = rgb_bandlist[0].stem
+        rio_meta["band_2"] = rgb_bandlist[1].stem
+        rio_meta["band_3"] = rgb_bandlist[2].stem
+        rio_meta["dtype"] = rgb_im.dtype.__str__()  # this should np.unit8
```
Comment (sixy6e) [inline sungc/deglint.py]:
```suggestion
        rio_meta["dtype"] = rgb_im.dtype.name  # this should np.unit8
```

File: sungc/deglint.py [hunk 5]
```diff
@@ -479,55 +580,61 @@ def nir_subtraction(self, vis_band_ids, nir_band_id, odir=None, waterVal=5):
...
+            with rasterio.open(vis_bandpath, "r") as ds_vis:
+                vis_im = ds_vis.read(1)
 
                 # get metadata and load NIR array
-                kwargs = dsVIS.meta.copy()
+                kwargs = ds_vis.meta.copy()
                 nodata = kwargs["nodata"]
-                spatialRes = int(abs(kwargs["transform"].a))
+
+                rio_funcs.check_image_singleval(vis_im, int(nodata), "vis_im")
```
Comment (sixy6e) [inline sungc/deglint.py]:
```suggestion
                rio_funcs.check_image_singleval(vis_im, nodata, "vis_im")
```

Comment (sixy6e) [reply sungc/deglint.py]: I'd not cast the nodata to int.  Your function supports `Union[float, int]`, could be expanded to `Union[float, int, None]`.
The case being _what if the type is None?_  You'll get an error forcing a None to an int.

File: sungc/deglint.py [hunk 6]
```diff
@@ -661,58 +768,64 @@ def hedley_2005(
...
+                vis_im = ds_vis.read(1)
 
                 # get metadata and load NIR array
-                kwargs = dsVIS.meta.copy()
+                kwargs = ds_vis.meta.copy()
                 nodata = kwargs["nodata"]
-                spatialRes = int(abs(kwargs["transform"].a))
+                spatial_res = int(abs(kwargs["transform"].a))
+
+                rio_funcs.check_image_singleval(vis_im, int(nodata), "vis_im")
```
Comment (sixy6e) [inline sungc/deglint.py]:
```suggestion
                rio_funcs.check_image_singleval(vis_im, nodata, "vis_im")
```

File: sungc/deglint.py [hunk 7]
```diff
@@ -918,42 +1036,54 @@ def cox_munk(
...
 
                 # get metadata
-                kwargs = dsVIS.meta.copy()
-                nodata = kwargs["nodata"]
-                spatialRes = int(abs(kwargs["transform"].a))
+                kwargs = ds_vis.meta.copy()
+                nodata = kwargs["nodata"]  # this is -999
+                spatial_res = int(abs(kwargs["transform"].a))
+
+                rio_funcs.check_image_singleval(vis_im, int(nodata), "vis_im")
```
Comment (sixy6e) [inline sungc/deglint.py]:
```suggestion
                rio_funcs.check_image_singleval(vis_im, nodata, "vis_im")
```

File: sungc/interactive.py [hunk 8]
```diff
@@ -43,12 +47,35 @@ class ROI_Selector(object):
           line connecting two existing vertices
     """
 
-    def __init__(self, ax=None, poly_xy=None, max_ds=10):
+    def __init__(
+        self,
+        ax: matplotlib.axes.Axes,
+        poly_xy: Union[list, tuple, None] = None,
```
Comment (sixy6e) [inline sungc/interactive.py]: Have a look at [mypy docs](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html#built-in-types) on how to use typing hints.

Comment (sixy6e) [reply sungc/interactive.py]: Might be easiest to look at a future PR to implement typing hints.

File: sungc/rasterio_funcs.py [hunk 9]
```diff
@@ -1,38 +1,55 @@
...
+
+    else:
+        if np.all(image == value):
+            raise Exception(f"{img_name} only contains a single value ({value})")
+
+
+def get_basename(filename: Path):
     """ Get a file's basename without extension """
-    return os.path.basename(os.path.splitext(filename)[0])
+    return filename.stem
```
Comment (sixy6e) [inline sungc/rasterio_funcs.py]: A superfluos function now as the input is a `Path` object, and where only calling a property on the `Path` object, namely `.stem`

File: sungc/rasterio_funcs.py [hunk 10]
```diff
@@ -70,22 +87,25 @@ def load_bands(band_list, scale_factor, apply_scaling):
...
-    nBands = len(band_list)
+    if scale_factor <= 0:
+        raise Exception("load_bands: scale_factor <= 0")
+
+    nbands = len(bandlist)
     spectral_cube = None
     nodata_val = None
     meta_dict = None
-    data_type = np.float32
+    data_type = np.dtype(np.float32)
```
Comment (sixy6e) [inline sungc/rasterio_funcs.py:94]: Not sure what this is doing ... `np.float32` will be of type `np.dtype`

File: sungc/rasterio_funcs.py [hunk 11]
```diff
@@ -98,27 +118,33 @@ def load_bands(band_list, scale_factor, apply_scaling):
             band[band <= 0] = nodata_val
 
             spectral_cube[z, :, :] = band
-            meta_dict["band_{0}".format(z + 1)] = get_basename(band_list[z])
+            meta_dict["band_{0}".format(z + 1)] = get_basename(bandlist[z])
 
-    meta_dict["count"] = nBands
+    meta_dict["count"] = nbands
+    meta_dict["dtype"] = data_type.__str__()
```
Comment (sixy6e) [inline sungc/rasterio_funcs.py]:
```suggestion
    meta_dict["dtype"] = data_type.name
```

File: sungc/rasterio_funcs.py (hunk 10 above)
Comment (Oceancolour-RG) [reply sungc/rasterio_funcs.py:94]:
```
dt1 = np.float32  # <class 'numpy.float32'>
dt1.name
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: type object 'numpy.float32' has no attribute 'name'
```
However,
```
dt2 = np.dtype(np.float32)
print(dt2.name)
float32
```

File: sungc/deglint.py (hunk 3 above)
Comment (Oceancolour-RG) [reply sungc/deglint.py]: You're right, it's meant to be "azimuthal-exiting"

File: sungc/deglint.py (hunk 4 above)
Comment (Oceancolour-RG) [reply sungc/deglint.py]: fixed

File: sungc/rasterio_funcs.py (hunk 10 above)
Comment (sixy6e) [reply sungc/rasterio_funcs.py:94]: Woops my bad.  Wasn't thinking straight.

File: sungc/deglint.py (hunk 5 above)
Comment (Oceancolour-RG) [reply sungc/deglint.py]: The issue is that nodata is typically a float, but the dtype of the bands are np.int16. If I don't change it to int() would the following still work?
`if np.all(image == value):`
where image is np.int16, while value is np.float64

Also, if `nodata = None`, then I wouldn't be able to run a check, and it possibly indicates a faulty geotiff

File: sungc/rasterio_funcs.py (hunk 9 above)
Comment (Oceancolour-RG) [reply sungc/rasterio_funcs.py]: This function was removed

File: sungc/rasterio_funcs.py (hunk 11 above)
Comment (Oceancolour-RG) [reply sungc/rasterio_funcs.py]: changed

File: sungc/deglint.py (hunk 5 above)
Comment (sixy6e) [reply sungc/deglint.py]: Yeah Python can compare an int array with a float value.

```python
>>> a = numpy.random.randint(0, 256, (3, 3)).astype('uint8')
>>> a
array([[ 25, 116,  60],
       [ 16, 195, 177],
       [195,  72,  88]], dtype=uint8)
>>> a == 25.0
array([[ True, False, False],
       [False, False, False],
       [False, False, False]])
```

The default return of any dataset opened by rasterio is `None` if there isn't a `nodata` field set.  So if `None` is the return nodata val, then I'd say skip the single value test.

**Review by sixy6e — APPROVED** [review]
Looks good

:+1:

**Pull request merged**